pysph/solver/tools.py

```diff
@@ -81,9 +81,51 @@ def __init__(self, app, array_name, props, freq=100, xi=None, yi=None,
         )
 
     def post_step(self, solver):
-        if solver.count%self.freq == 0 and solver.count > 0:
+        if solver.count % self.freq == 0 and solver.count > 0:
             self.interp.nnps.update()
             data = dict(x=self.xi, y=self.yi, z=self.zi)
             for prop in self.props:
                 data[prop] = self.interp.interpolate(prop)
             self.array.set(**data)
+
+
+class DensityCorrection(Tool):
+    """
+    A tool to reinitialize the density of the fluid particles
+    """
+
+    def __init__(self, app, m=5, corr=0, arr_ind=[0], dim=2, kernel=None):
```

Comment:
- Maybe you should just ask for the names of the specific particle arrays that the user wants to apply the correction for and not the indices.  
- Instead of corr and m, instead use freq or something better and if that is zero don't apply the correction.
- Always insist on the value of dim, or otherwise automatically figure it out by looking at the solver's dim.

Reply:
You may want to add the property `tw` when you get the arrays and maybe use something else for the name.